File: backend/lib/mihomo.py
import httpx
import json

MIHOMO_BASE_URL = "https://api.mihomo.me/sr_info_parsed"

async def fetch_mihomo_data(uid: str, lang: str = "en") -> dict | None:
    """
    Fetches raw parsed data from Mihomo API for the given UID.
    """
    url = f"{MIHOMO_BASE_URL}/{uid}?lang={lang}"
    headers = {
        "User-Agent": "OmniApp/3.0",
        "Accept": "application/json"
    }
    
    async with httpx.AsyncClient() as client:
        try:
            response = await client.get(url, headers=headers, timeout=10.0)
            if response.status_code == 200:
                return response.json()
            else:
                print(f"Mihomo API Error {response.status_code}: {response.text}")
                return None
        except Exception as e:
            print(f"Network error fetching Mihomo data: {e}")
            return None

def clean_mihomo_data(raw_data: dict) -> dict:
    """
    Cleans and formats the raw data from Mihomo into a structured profile object.
    Calculates Relic Crit Value (CV) and assigns a Grade.
    """
    if not raw_data or "player" not in raw_data:
        return {}

    player = raw_data["player"]
    characters_raw = raw_data.get("characters", [])
    
    cleaned_characters = []
    for char in characters_raw:
        cleaned_char = {
            "id": char.get("id"),
            "name": char.get("name"),
            "rarity": char.get("rarity"),
            "level": char.get("level"),
            "element": char.get("element", {}).get("name"),
            "path": char.get("path", {}).get("name"),
            "icon": char.get("icon"),
            "preview": char.get("preview"),
            "portrait": char.get("portrait"),
            "rank": char.get("rank", 0),
            "light_cone": char.get("light_cone"),
            "properties": char.get("properties", []),
            "skills": char.get("skills", []),
            "relics": []
        }
        
        # Extract relics and calculate CV
        for relic in char.get("relics", []):
            sub_stats = relic.get("sub_affix", [])
            
            # Calculate CV
            cv = 0.0
            for sub in sub_stats:
                field = sub.get("field", "")
                val = sub.get("value", 0.0)
                if field == "crit_rate":
                    cv += (val * 100) * 2
                elif field == "crit_dmg":
                    cv += (val * 100)
                    
            # Assign Grade
            grade = "C"
            if cv >= 35:
                grade = "S"
            elif cv >= 25:
                grade = "A"
            elif cv >= 15:
                grade = "B"
                
            cleaned_relic = {
                "id": relic.get("id"),
                "name": relic.get("name"),
                "set_name": relic.get("set_name"),
                "rarity": relic.get("rarity"),
                "level": relic.get("level"),
                "icon": relic.get("icon"),
                "main_stat": relic.get("main_affix", {}),
                "sub_stats": sub_stats,
                "score": {
                    "cv": round(cv, 1),
                    "grade": grade
                }
            }
            cleaned_char["relics"].append(cleaned_relic)
            
        cleaned_characters.append(cleaned_char)
        
    return {
        "uid": player.get("uid"),
        "nickname": player.get("nickname"),
        "level": player.get("level"),
        "signature": player.get("signature"),
        "avatar": player.get("avatar", {}).get("icon"),
        "space_info": player.get("space_info", {}),
        "characters": cleaned_characters
    }

File: frontend/src/lib/mihomo.ts
interface RelicSubStat {
  type: string;
  field: string;
  name: string;
  value: number;
  display: string;
  percent: boolean;
}

// Extracted Relic interface if needed elsewhere

export interface CleanedProfile {
  uid: string;
  nickname: string;
  level: number;
  avatarId: string;
  signature: string;
  characters: any[]; // Extended type can be added later
}

/**
 * Fetches profile data from Mihomo API and cleans it to remove unnecessary fields.
 */
export async function fetchAndCleanProfile(uid: string): Promise<CleanedProfile> {
  const response = await fetch(`https://api.mihomo.me/sr_info_parsed/${uid}?lang=id`);
  
  if (!response.ok) {
    if (response.status === 404 || response.status === 400) {
      throw new Error('UID not found');
    }
    throw new Error('Upstream error from Mihomo API');
  }
  
  const raw = await response.json();
  
  // Clean the payload
  const cleaned: CleanedProfile = {
    uid: raw.player?.uid,
    nickname: raw.player?.nickname,
    level: raw.player?.level,
    avatarId: raw.player?.avatar?.id,
    signature: raw.player?.signature,
    characters: raw.characters?.map((c: any) => ({
      id: c.id,
      name: c.name,
      rarity: c.rarity,
      level: c.level,
      promotion: c.promotion,
      path: {
        id: c.path?.id,
        name: c.path?.name
      },
      element: {
        id: c.element?.id,
        name: c.element?.name
      },
      attributes: c.attributes,
      additions: c.additions,
      relics: c.relics?.map((r: any) => ({
        id: r.id,
        name: r.name,
        setId: r.set_id,
        setName: r.set_name,
        rarity: r.rarity,
        level: r.level,
        mainStat: {
          name: r.main_affix?.name,
          value: r.main_affix?.value,
          display: r.main_affix?.display,
          percent: r.main_affix?.percent,
        },
        subStats: r.sub_affix?.map((sub: any) => ({
          type: sub.type,
          field: sub.field,
          name: sub.name,
          value: sub.value,
          display: sub.display,
          percent: sub.percent,
        })) || []
      })),
      lightCone: c.light_cone ? {
        id: c.light_cone.id,
        name: c.light_cone.name,
        rarity: c.light_cone.rarity,
        level: c.light_cone.level,
        attributes: c.light_cone.attributes
      } : null
    })) || []
  };

  return cleaned;
}
